server/src/main/java/org/elasticsearch/telemetry/Trace.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.telemetry;
+
+import org.elasticsearch.common.UUIDs;
+import org.elasticsearch.core.Releasable;
+import org.elasticsearch.telemetry.tracing.Traceable;
+import org.elasticsearch.telemetry.tracing.Tracer;
+import org.elasticsearch.threadpool.ThreadPool;
+
+import java.util.Map;
+
+public class Trace {
+
+    public static Releasable trace(ThreadPool threadPool, Tracer tracer, String name) {
+        return trace(threadPool, tracer, name, Map.of());
+    }
+
+    /**
+     * Creates a new trace for a synchronous block of code.
+     * @return a span that needs to be released once block of code is completed
+     */
+    public static Releasable trace(ThreadPool threadPool, Tracer tracer, String name, Map<String, Object> attributes) {
+        if (tracer.isEnabled() == false) {
+            return () -> {};
+        }
+        var span = Span.create();
+        var ctx = threadPool.getThreadContext().newTraceContext();
+        tracer.startTrace(threadPool.getThreadContext(), span, name, attributes);
```

Comment:
First off: this is great and allows to create sub-spans of the current thread context very easily.

However, I wonder if we should use `threadPool.getThreadContext().newTraceContext()`; this mutates the current thread local state, and in particular only works if there's a strict hierarchy of spans. @elastic/es-core-infra what do you think?

It might be sufficient to copy the logic from `newTraceContext` to create a new, separate `TraceContext` and store it as part of the `Releaseable` that is returned here. I attempted that [here ](https://github.com/elastic/elasticsearch/pull/109528/files#diff-a201f5c75a4783e2e9becba7ba56b7799d1425747c42537f42be53d2456a377bR133-R13) and it _seemed_ to work fine.

This would allow create interleaved traces in the same thread which would provide more flexibility; my use case for this is seeing when individual ESQL operators start and stop, even though they are executed concurrently on the same thread.

Reply:
I am worried about the use of threadpool & threadcontext here.
 we should aim to use org.eleasticsearch.telemetry.tracing.TraceContext 
I think it should be possible to either:
- refactor the Pr to have the new context and context to restore (would be probably a bit ugly)
- refactor the TraceContext to create a newTraceContext

Reply:
>refactor the Pr to have the new context and context to restore (would be probably a bit ugly)

Not sure I follow, `threadPool.getThreadContext().newTraceContext()` creates a new context and returns a releasable to rollback to the previous one. 
To me it sounds like what is suggested, however implicit (may be the comment is about it?)

Reply:
> refactor the Pr to have the new context and context to restore (would be probably a bit ugly)

as part of a method signature.
My point is that we should not try to add ThreadContext in the telemetry api